opm/simulators/wells/MultisegmentWellEquations.cpp

```diff
@@ -0,0 +1,397 @@
+        // the number of the row corrspnds to the segment now
+        const int seg = row.index();
+        // adding the item related to outlet relation
+        const Segment& segment = well_.segmentSet()[seg];
+        const int outlet_segment_number = segment.outletSegment();
+        if (outlet_segment_number > 0) { // if there is a outlet_segment
+            const int outlet_segment_index = well_.segmentNumberToIndex(outlet_segment_number);
+            row.insert(outlet_segment_index);
+        }
+
+        // Add nonzeros for diagonal
+        row.insert(seg);
+
+        // insert the item related to its inlets
+        for (const int& inlet : well_.segmentInlets()[seg]) {
+            row.insert(inlet);
+        }
+    }
+
+    // make the C matrix
+    for (auto row = duneC_.createbegin(),
+              end = duneC_.createend(); row != end; ++row) {
+        // the number of the row corresponds to the segment number now.
+        for (const int& perf : well_.segmentPerforations()[row.index()]) {
+            const int cell_idx = cells[perf];
+            row.insert(cell_idx);
+        }
+    }
+
+    // make the B^T matrix
+    for (auto row = duneB_.createbegin(),
+              end = duneB_.createend(); row != end; ++row) {
+        // the number of the row corresponds to the segment number now.
+        for (const int& perf : well_.segmentPerforations()[row.index()]) {
+            const int cell_idx = cells[perf];
+            row.insert(cell_idx);
+        }
+    }
+
+    resWell_.resize(well_.numberOfSegments());
+}
+
+template<class Scalar, int numWellEq, int numEq>
+void MultisegmentWellEquations<Scalar,numWellEq,numEq>::clear()
+{
+    duneB_ = 0.0;
+    duneC_ = 0.0;
+    duneD_ = 0.0;
+    resWell_ = 0.0;
+    duneDSolver_.reset();
+}
+
+template<class Scalar, int numWellEq, int numEq>
+void MultisegmentWellEquations<Scalar,numWellEq,numEq>::
+apply(const BVector& x, BVector& Ax) const
+{
+    BVectorWell Bx(duneB_.N());
+
+    duneB_.mv(x, Bx);
+
+    // invDBx = duneD^-1 * Bx_
+    const BVectorWell invDBx = mswellhelpers::applyUMFPack(*duneDSolver_, Bx);
+
+    // Ax = Ax - duneC_^T * invDBx
+    duneC_.mmtv(invDBx,Ax);
+}
+
+template<class Scalar, int numWellEq, int numEq>
+void MultisegmentWellEquations<Scalar,numWellEq,numEq>::
+apply(BVector& r) const
+{
+    // invDrw_ = duneD^-1 * resWell_
+    const BVectorWell invDrw = mswellhelpers::applyUMFPack(*duneDSolver_, resWell_);
+    // r = r - duneC_^T * invDrw
+    duneC_.mmtv(invDrw, r);
+}
+
+template<class Scalar, int numWellEq, int numEq>
+void MultisegmentWellEquations<Scalar,numWellEq,numEq>::createSolver()
+{
+#if HAVE_UMFPACK
+    if (duneDSolver_) {
+        return;
+    }
+
+    duneDSolver_ = std::make_shared<Dune::UMFPack<DiagMatWell>>(duneD_, 0);
+#else
+    OPM_THROW(std::runtime_error, "MultisegmentWell support requires UMFPACK."
+              "Reconfigure opm-simulators with SuiteSparse/UMFPACK support and recompile.");
```

Comment:
I take it this message is unchanged from its original form, but maybe we can use this opportunity to add a space after the full stop.  As or right now, the diagnostic message is
```
... requires UMFPACK.Reconfigure opm-simulators ...
```
and that is less than ideal in my opinion.